library/ecdsa.c

```diff
@@ -540,7 +540,7 @@ static int ecdsa_verify_restartable( mbedtls_ecp_group *grp,
     if( mbedtls_mpi_cmp_int( r, 1 ) < 0 || mbedtls_mpi_cmp_mpi( r, &grp->N ) >= 0 ||
         mbedtls_mpi_cmp_int( s, 1 ) < 0 || mbedtls_mpi_cmp_mpi( s, &grp->N ) >= 0 )
     {
-        ret = MBEDTLS_ERR_ECP_VERIFY_FAILED;
+        ret = MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
```

Comment:
`ecdsa.h` says
```
 * \return          #MBEDTLS_ERR_ECP_BAD_INPUT_DATA if the signature
 *                  is invalid.
```

so that's the error code I'd expect in this situation.

Reply:
But the description of `MBEDTLS_ERR_ECP_VERIFY_FAILED` is “the signature is not valid”...

I think this was a mistake in the documentation of `mbedtls_ecdsa_verify`.

In any case, I don't see any basis to separate between the case where one of r or s is out of range for the given curve, and the case where r and s are in range but do not form a valid signature for the particular key.

Reply:
OK, my reading was that we were trying to distinguish between an invalid (e.g., corrupted) signature, vs one that did not match. Maybe that is not a useful distinction?

Reply:
It could be a useful distinction in some contexts, but the distinction is not clear-cut. For example `mbedtls_ecdsa_read_signature` returns `MBEDTLS_ERR_ECP_BAD_INPUT_DATA` if the signature doesn't have the expected ASN.1 structure, `MBEDTLS_ERR_ECP_VERIFY_FAILED` if the signature can be parsed into two numbers but those numbers aren't good, and 0 if the signature is correct. I think that makes a little more sense (for a low-level API) than putting the threshold between “either can't be parsed or can be parsed into numbers that aren't in range” and “can be parsed into numbers that are in range but aren't correct for the key”. But it's something we really should document, and it's probably not what I'd choose if I was designing this API from scratch.